x-pack/plugins/task_manager/server/task_validator.ts

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { max, memoize } from 'lodash';
+import type { Logger } from '@kbn/core/server';
+import type { ObjectType } from '@kbn/config-schema';
+import { TaskTypeDictionary } from './task_type_dictionary';
+import type { TaskInstance, ConcreteTaskInstance, TaskDefinition } from './task';
+
+interface TaskValidatorOpts {
+  allowReadingInvalidState: boolean;
+  definitions: TaskTypeDictionary;
+  logger: Logger;
+}
+
+type LatestStateSchema =
+  | undefined
+  | {
+      schema: ObjectType;
+      version: number;
+      up: (state: Record<string, unknown>) => Record<string, unknown>;
+    };
+
+export class TaskValidator {
+  private readonly logger: Logger;
+  private readonly definitions: TaskTypeDictionary;
+  private readonly allowReadingInvalidState: boolean;
+  private readonly cachedGetLatestStateSchema: (taskTypeDef: TaskDefinition) => LatestStateSchema;
+  private readonly cachedExtendSchema: typeof extendSchema;
+
+  constructor({ definitions, allowReadingInvalidState, logger }: TaskValidatorOpts) {
+    this.logger = logger;
+    this.definitions = definitions;
+    this.allowReadingInvalidState = allowReadingInvalidState;
+    this.cachedGetLatestStateSchema = memoize(
+      getLatestStateSchema,
+      (taskTypeDef) => taskTypeDef.type
+    );
+    this.cachedExtendSchema = memoize(
+      extendSchema,
+      // We need to cache two outcomes per task type (unknowns: ignore and unknowns: forbid)
+      (options) => `${options.taskType}|unknowns:${options.unknowns}`
+    );
+  }
+
+  public getValidatedTaskInstance<T extends TaskInstance>(
+    task: T,
+    mode: 'read' | 'write',
```

Comment:
Wouldn't separating read and write functions be better?

Reply:
I'll PoC it and see how the code compares, let me circle back!

Reply:
The PoC turned out great so I committed it here: https://github.com/elastic/kibana/pull/159048/commits/4ee1471d197d5d6acca6fa87fcc23367a22d5c8e